### ISSUE TITLE
ztex: working -ztex format blocks --list= options to retrieve format info

### ISSUE THREAD

**Issue by AlekseyCherepanov** [magnumripper/JohnTheRipper#3379]
While `--format=bcrypt-ztex` works, john on other terminal cannot show output for `--list=format-all-details` or `--list=format-details`. With explicit `--format=bcrypt-ztex` or `--format=descrypt-ztex`, `--list` fails too. But I can choose non-ztex formats well.

So cracking with `-ztex` format
1) blocks other `-ztex` formats from showing info with `--list=` options,
2) blocks showing info for all lists of formats that include `-ztex`.

### Steps to reproduce ###
Run some `-ztex` format in one window and try `--list=format-all-details` in other window.

(2 boards connected, both are cracking with bcrypt-ztex)
```
$ john --list=format-all-details
SN 0...: usb_claim_interface error -6 (LIBUSB_ERROR_BUSY)
SN 0...: usb_claim_interface error -6 (LIBUSB_ERROR_BUSY)
no valid ZTEX devices found
```

### System configuration ###
```
$ john --list=build-info
Version: 1.8.0.13-jumbo-1-bleeding-5d6ec13 2018-08-07 22:30:42 +0530
Build: linux-gnu 64-bit x86_64 SSE4.2 AC
[...]
gcc version: 5.1.0
GNU libc version: 2.19 (loaded: 2.19)
```

**Comment by solardiz** [comment]
I am not sure what can be done here. I guess initialization of ZTEX devices may be postponed or skipped for options that don't require use of those devices - such as when querying info on non-ZTEX formats. However, I guess info on ZTEX formats depends on the number of boards, and determining how many boards there are involves talking to them, which can't be done reliably when they're already in use.

**Comment by solardiz** [comment]
> But I can choose non-ztex formats well.

Oh. Then I think what we have here is intended behavior. With our support for multiple boards, we can't reasonably query ZTEX formats when they're already in use, or can we? Denis, if you see no easy way to improve upon the current situation, then please close this issue.

**Comment by frank-dittrich** [comment]
Is it possible to at least write some  information to stderr which explains to the user what the problem is?

**Comment by magnumripper** [comment]
If a user doesn't understand `LIBUSB_ERROR_BUSY` in this particular situation, I seriously doubt anything we add will help. But that's just me.

**Comment by AlekseyCherepanov** [comment]
Without connected boards, the output does not contain `LIBUSB_ERROR_BUSY`:
```
$ john --list=format-all-details
no valid ZTEX devices found
```

With connected idle boards, output from initialization breaks the table from `--list=format-details`:
```
$ john --list=format-details --format='*-ztex'
SN 0XXXXXXXXY: uploading bitstreams.. ok
SN 0XXXXXXXXZ: uploading bitstreams.. ok
ZTEX 0XXXXXXXXY bus:10 dev:8 Frequency:220,160 220,160 220,160 220,160 
ZTEX 0XXXXXXXXZ bus:2 dev:2 Frequency:220,160 220,160 220,160 220,160 
descrypt-ztex	8	1	1	00600001	9	DES ZTEX	traditional crypt(3)		0	5	2		0	CCNf8Sbh3HDfQ
SN 0XXXXXXXXY: uploading bitstreams.. ok
SN 0XXXXXXXXZ: uploading bitstreams.. ok
ZTEX 0XXXXXXXXY bus:10 dev:8 Frequency:146 146 146 146 
ZTEX 0XXXXXXXXZ bus:2 dev:2 Frequency:146 146 146 146 
bcrypt-ztex	72	1	1	00600003	9	Blowfish ZTEX			0	4	20	iteration count	0	$2a$05$CCCCCCCCCCCCCCCCCCCCC.VGOzA784oUp/Z0DY336zx7pLYAy0lwK
SN 0XXXXXXXXY: uploading bitstreams.. ok
SN 0XXXXXXXXZ: uploading bitstreams.. ok
ZTEX 0XXXXXXXXY bus:10 dev:8 Frequency:135 135 135 135 
ZTEX 0XXXXXXXXZ bus:2 dev:2 Frequency:135 135 135 135 
sha512crypt-ztex	64	1	1	00400003	14	sha512crypt ZTEX	crypt(3) $6$	(rounds=5000)	-1	4	24	iteration count	0	$6$LKO/Ute40T3FNF95$6S/6T2YuOIHY0N3XpLKABJ3soYcXD9mB7uVbtEZDj/LNscVhZoZ9DEH.sBciDrMsHOWOoASbNLTypH/5X26gN0
ZTEX 0XXXXXXXXY bus:10 dev:8 Frequency:135 135 135 135 
ZTEX 0XXXXXXXXZ bus:2 dev:2 Frequency:135 135 135 135 
Drupal7-ztex	47	1	1	00400003	3	SHA512 ZTEX	$S$	(x16385)	-1	4	9	iteration count	0	$S$CwkjgAKeSx2imSiN3SyBEg8e0sgE2QOx4a/VIfCHN0BZUNAWCr1X
```

Example of output for `--list=format-all-details`:
```
$ john --list=format-all-details --format='bcrypt-ztex'
SN 0XXXXXXXXY: uploading bitstreams.. ok
SN 0XXXXXXXXZ: uploading bitstreams.. ok
ZTEX 0XXXXXXXXY bus:10 dev:8 Frequency:146 146 146 146 
ZTEX 0XXXXXXXXZ bus:2 dev:2 Frequency:146 146 146 146 
Format label                         bcrypt-ztex
 Disabled in configuration file      no
Min. password length                 0
Max. password length                 24 [worst case UTF-8] to 72 [ASCII]
Min. keys per crypt                  1
Max. keys per crypt                  1
Flags
 Case sensitive                      yes
 Truncates at max. length            yes
 Supports 8-bit characters           yes
 Converts internally to UTF-16/UCS-2 no
 Honours --encoding=NAME             n/a
 Collisions possible (as in likely)  no
 Uses a bitslice implementation      no
 The split() method unifies case     no
 Supports very long hashes           no
 Internal mask generation            yes (device target: 0x)
 A $dynamic$ format                  no
 A dynamic sized salt                no
Number of test vectors               9
Algorithm name                       Blowfish ZTEX
Format name                          
Benchmark comment                    
Benchmark length                     0
Binary size                          4
Salt size                            20
Tunable cost parameters              iteration count
Example ciphertext                   $2a$05$CCCCCCCCCCCCCCCCCCCCC.VGOzA784oUp/Z0DY336zx7pLYAy0lwK
```

**Comment by solardiz** [comment]
Oh, I thought this would report the correct min/max_keys_per_crypt for the actual number of boards, but it doesn't even report them for one board (which could be done without actually talking to the board). Fixing that would be a more important improvement than allowing to query the format parameters when the boards are in use.

**Comment by magnumripper** [comment]
Actually it **may** be that the call to `init()` stems from long ago when OpenCL formats set min/max kpc in that function - but they too do it in `reset()` nowadays. I need to look around a bit too. I'll report back.

**Comment by magnumripper** [comment]
OK, so format's `init()` is called for four --list= functions:

`--list=format-methods`, because a format may set or change some function pointers in init().

`--list=format-tests`, because some formats change some of the test vectors depending on encoding.

`--list=format-details` and `--list=format-all-details`, mainly because some formats change their max plaintext length when encoding is used, or KPC when under OpenMP. Ideally we'd show a relevant KPC for OpenCL/ZTEX formats as well but it will vary with auto-tune anyway (at least for OpenCL) and it takes too long to auto-tune for --list.

I pushed a hack disabling the call to `init()` for ZTEX formats. Please close this issue if that fixed it and didn't introduce some other problem.

**Comment by Apingis** [comment]
I've tested, looks OK with boards on and off.
A minor glitch is in ```--list=format-all-details --form=<name>-ztex```:
``` Internal mask generation            yes (device target: 0x)```
If mask is specified on command line it's the same. It sets mask's target in ```init()```.

Another point by @solardiz (if I got it right) is that from UX perspective, a user would want to see
substantial information regarding KPC, e.g. how many keys per board are processed in parallel for given format. Questions may arise such as:
- Is ```--list=format-all-details``` a debugging / development feature or it's planned as meaningful for end user? E.g. a user has "heavy" hashes and wants to know KPC per board, to arrange the wordlist. Such as the user might want to check a big number of "heavy" hashes vs. relatively small "most probable" wordlist and the user wants exact number (e.g. 640 per board for sha512crypt-ztex).

I think that's beyond the scope of this issue, I'm closing it.

**Comment by magnumripper** [comment]
> Another point by @solardiz (if I got it right) is that from UX perspective, a user would want to see
substantial information regarding KPC, e.g. how many keys per board are processed in parallel for given format

I'd love that, but the devil is in the details. One problem of several is that the KPC not only depends on auto-tune being run first, but that the outcome of such would depend on what was loaded (eg. iteration count). So to get a figure for a specific case, you'd need to supply a hash file as well.

At some point in time I think we did report KPC for OpenCL - but that was after a substantial delay of auto-tuning (which became very problematic in case you'd like to see data for ALL formats and not just a single one) and that tuning was made with whatever *test vectors* happened to be picked. Sometimes the figure wouldn't be anywhere near of reflecting your actual hashes, which could be several orders of magnitude easier or harder to attack.

**Comment by Apingis** [comment]
> One problem of several is that the KPC not only depends on auto-tune being run first, but that the 
> outcome of such would depend on what was loaded (eg. iteration count). 

Well, I understand your points, there's however 1 thing worth mentioning: the number of keys processed in parallel, per device. In ztex formats that's well known long before compile time. Actual KPC depends on factors you mentioned, still it's a multiple or equal to the number of keys processed by the device in parallel, multiplied by the number of devices. That's what end user might want to know.

I'm not proficient with OpenCL, just a random thought: can we know a number of cores in OpenCL device out of its runtime, deduct the number of keys processed in parrallel (mb. multiply by 2 or 3 if there's some "interleaved" implementation). However I'm unsure what to do if there're several unequal devices.

Another directon of thought is: if we'd want / plan some user-friendly command-line option be more descriptive about given format (maybe base on existing command-line option), I'd proposed the following idea:
- In ```struct fmt_main``` create additional function pointer that would provide per-format (or per group of formats) description on stdout. For ztex formats that would be base KPC (number of keys computed in parallel), and human readable description on its dependance on TargetSetting / TargetRounds.

Of course the user can always ```grep``` through the sources and see the number in ```struct device_bitstream.min_keys```.

**Comment by solardiz** [comment]
Aleksey, it sounds like what you ask for is described in issue #3386. You could want to contribute to addressing that one. ;-)

@Apingis good suggestions, thanks! And here's mine for you to consider implementing: maybe you could set the initial (seen before any function call) min/max kpc to the absolute minimums needed to load all cores on one board, which are known at compile-time. For example, for bcrypt-ztex you'd set them to 496. This would be of some help already, compared to the current settings of 1/1, which are purely dummy. You could also adjust the reported max_keys_per_crypt based on `Target*` if you can parse it in time (from `reset`?) You wouldn't factor in the board count, because that requires talking to the boards and thus has usability drawbacks. Ideally, we'd print a warning about the board count not being factored into this reporting - can do so after a `strstr` for `-ztex` in shared code (would be a hack, yet would help the users).

**Comment by Apingis** [comment]
ztex formats don't enable onboard generator starting from some cost / rounds as in such cases it would be  performance degradation. Example: a user specifies mask '?l?l' (676 candidates per base word).
bcrypt-ztex at cost 10 determines KPC to be 496x4 = 1984 candidates/board. If on-board generator was enabled, then it would be 2 or 3 base words (internally sometimes referred to as "template keys") per ```crypt_all()```, thats 676x2=1352 or 676x3=2028 candidates. In both cases that would result in some cores being idle, resulting in substantial loss of performance: there're 124 bcrypt cores in fpga, 496/board, to utilize resources fully KPC must be this number or a multiple of it.
I'm not sure how do OpenCL formats with on-device generator behave on "slow" hashes. The approach might be to disable on-board generation starting from some cost/rounds.
The essense is that there's an important for slow hashes parameter, the "number of hashes processed by the device in parallel", which might be overlooked in JtR development and it may substantially affect performance on slow hashes.
This relates to #3386.
Consider a "corner case": a bcrypt hash with setting 17 (takes ~15 sec to process 496 hashes/board, requires to adjust jtr_device.c:DEVICE_TIMEOUT). If 495 candidates are enqueued for processing there wouldn't be a big loss and if 497 are enqueued then performance loss would be near 50%.

> BTW bcrypt-ztex tells the following for cost 10 too. I think --mask does not improve speed for cost 10.
> Note: This format may be a lot faster with --mask acceleration (see doc/MASK).

Yeah, for cost 10 the message is completely irrelevant.

**Comment by Apingis** [comment]
@solardiz I'll set min/max keys_per_crypt in ```struct fmt_main``` to reflect number of keys processed by the device in parallel, for the display in ```--list=format-all-details``` (that would be overwritten on a real run anyway).
My idea in line with improvements for user friendliness is: use ```options.verbosity``` facility for more reporting/advising, I've already started development in that directon but didn't commit anything yet.

**Comment by solardiz** [comment]
@Apingis All of what you say here makes sense to me, and that's a good plan. Thanks.

We don't have on-device mask for any slow hashes implemented in OpenCL. We only have on-device mask for fast hashes, and in all ZTEX formats because of slower and shared communication to the device(s) (shared USB 2.0 hub vs. dedicated PCIe lanes) and also because you started with descrypt-ztex, which is a fast hash. (Don't get me wrong - it's great that you also implemented on-device mask for the slower hashes on ZTEX. This does help with larger clusters and/or lower hash cost settings. Thank you.)

Regarding `DEVICE_TIMEOUT`, maybe you could have it adjusted automatically when `TargetSetting` is large enough to require that? It'd be even better to have it adjusted based on actual highest cost setting in the loaded hashes, but that's more work.
